src/infra/dex/balancer/v3.rs

```diff
@@ -0,0 +1,169 @@
+//! BatchRouter. It serves as a thin wrapper around the `ethcontract` generated
+//! bindings, defining structs with named fields instead of using tuples.
+
+use {
+    crate::domain::{dex, eth},
+    contracts::{ethcontract::Bytes, BalancerV3BatchRouter, Permit2 as Permit2Contract},
+    ethereum_types::{H160, U256},
+};
+
+pub struct Permit2(Permit2Contract);
+
+impl Permit2 {
+    pub fn new(address: eth::ContractAddress) -> Self {
+        Self(contracts::dummy_contract!(Permit2Contract, address.0))
+    }
+
+    pub fn address(&self) -> eth::ContractAddress {
+        eth::ContractAddress(self.0.address())
+    }
+}
+
+pub struct Router(BalancerV3BatchRouter);
+
+pub struct SwapPathStep {
+    pub pool: H160,
+    pub token_out: H160,
+    /// If true, the "pool" is an ERC4626 Buffer. Used to wrap/unwrap tokens if
+    /// pool doesn't have enough liquidity.
+    pub is_buffer: bool,
+}
+
+pub struct SwapPath {
+    pub token_in: H160,
+    pub steps: Vec<SwapPathStep>,
+    pub input_amount_raw: U256,
+    pub output_amount_raw: U256,
+}
+
+impl Router {
+    pub fn new(address: eth::ContractAddress) -> Self {
+        Self(contracts::dummy_contract!(BalancerV3BatchRouter, address.0))
+    }
+
+    pub fn address(&self) -> eth::ContractAddress {
+        eth::ContractAddress(self.0.address())
+    }
+
+    pub fn swap_exact_amount_in(
+        &self,
+        paths: Vec<SwapPath>,
+        permit2: &Permit2,
+        v3_batch_router: &Router,
+        token_in: H160,
+        max_input: U256,
+    ) -> Vec<dex::Call> {
+        let permit2_approval_call =
+            Self::permit2_approval(permit2, v3_batch_router, token_in, max_input);
+
+        let swap_call = self.0.swap_exact_in(
+            Self::encode_paths(paths),
+            Self::deadline(),
+            Self::weth_is_eth(),
+            Self::user_data(),
+        );
+
+        vec![
+            permit2_approval_call,
+            dex::Call {
+                to: self.address(),
+                calldata: swap_call.tx.data.unwrap().0,
+            },
+        ]
+    }
+
+    pub fn swap_exact_amount_out(
+        &self,
+        paths: Vec<SwapPath>,
+        permit2: &Permit2,
+        v3_batch_router: &Router,
+        token_in: H160,
+        max_input: U256,
+    ) -> Vec<dex::Call> {
+        let permit2_approval_call =
+            Self::permit2_approval(permit2, v3_batch_router, token_in, max_input);
+
+        let swap_call = self.0.swap_exact_out(
+            Self::encode_paths(paths),
+            Self::deadline(),
+            Self::weth_is_eth(),
+            Self::user_data(),
+        );
+
+        vec![
+            permit2_approval_call,
+            dex::Call {
+                to: self.address(),
+                calldata: swap_call.tx.data.unwrap().0,
+            },
+        ]
+    }
+
+    // Transfers are done via Permit2, so we approve the balancer v3 router to spend
+    // the input tokens
+    fn permit2_approval(
```

Comment:
Seems like this function would better be defined on the `Permit2` struct instead.